x/wasm/keeper/handler_plugin_encoders_test.go

```diff
@@ -559,24 +629,168 @@ func TestEncoding(t *testing.T) {
+						ProposalId: 1,
+						Options: []wasmvmtypes.WeightedVoteOption{
+							{Option: wasmvmtypes.Yes, Weight: "0.23"},
+							{Option: wasmvmtypes.No, Weight: "0.24"},
+							{Option: wasmvmtypes.Abstain, Weight: "0.26"},
+							{Option: wasmvmtypes.NoWithVeto, Weight: "0.27"},
+						},
+					},
+				},
+			},
+			output: []sdk.Msg{
+				&v1.MsgVoteWeighted{
+					ProposalId: 1,
+					Voter:      myAddr.String(),
+					Options: []*v1.WeightedVoteOption{
+						{Option: v1.OptionYes, Weight: sdk.NewDecWithPrec(23, 2).String()},
+						{Option: v1.OptionNo, Weight: sdk.NewDecWithPrec(24, 2).String()},
+						{Option: v1.OptionAbstain, Weight: sdk.NewDecWithPrec(26, 2).String()},
+						{Option: v1.OptionNoWithVeto, Weight: sdk.NewDecWithPrec(27, 2).String()},
+					},
+				},
+			},
+		},
+		"Gov weighted vote: duplicate option": {
+			sender: myAddr,
+			srcMsg: wasmvmtypes.CosmosMsg{
+				Gov: &wasmvmtypes.GovMsg{
+					VoteWeighted: &wasmvmtypes.VoteWeightedMsg{
+						ProposalId: 1,
+						Options: []wasmvmtypes.WeightedVoteOption{
+							{Option: wasmvmtypes.Yes, Weight: "0.5"},
+							{Option: wasmvmtypes.Yes, Weight: "0.5"},
+						},
+					},
+				},
+			},
+			output: []sdk.Msg{
+				&v1.MsgVoteWeighted{
+					ProposalId: 1,
+					Voter:      myAddr.String(),
+					Options: []*v1.WeightedVoteOption{
+						{Option: v1.OptionYes, Weight: sdk.NewDecWithPrec(5, 1).String()},
+						{Option: v1.OptionYes, Weight: sdk.NewDecWithPrec(5, 1).String()},
+					},
+				},
+			},
+			expInvalid: true,
+		},
+		"Gov weighted vote: weight sum exceeds 1": {
+			sender: myAddr,
+			srcMsg: wasmvmtypes.CosmosMsg{
+				Gov: &wasmvmtypes.GovMsg{
+					VoteWeighted: &wasmvmtypes.VoteWeightedMsg{
+						ProposalId: 1,
+						Options: []wasmvmtypes.WeightedVoteOption{
+							{Option: wasmvmtypes.Yes, Weight: "0.51"},
+							{Option: wasmvmtypes.No, Weight: "0.5"},
+						},
+					},
+				},
+			},
+			output: []sdk.Msg{
+				&v1.MsgVoteWeighted{
+					ProposalId: 1,
+					Voter:      myAddr.String(),
+					Options: []*v1.WeightedVoteOption{
+						{Option: v1.OptionYes, Weight: sdk.NewDecWithPrec(51, 2).String()},
+						{Option: v1.OptionNo, Weight: sdk.NewDecWithPrec(5, 1).String()},
+					},
+				},
+			},
+			expInvalid: true,
+		},
+		"Gov weighted vote: weight sum less than 1": {
+			sender: myAddr,
+			srcMsg: wasmvmtypes.CosmosMsg{
+				Gov: &wasmvmtypes.GovMsg{
+					VoteWeighted: &wasmvmtypes.VoteWeightedMsg{
+						ProposalId: 1,
+						Options: []wasmvmtypes.WeightedVoteOption{
+							{Option: wasmvmtypes.Yes, Weight: "0.49"},
+							{Option: wasmvmtypes.No, Weight: "0.5"},
+						},
+					},
+				},
+			},
+			output: []sdk.Msg{
+				&v1.MsgVoteWeighted{
+					ProposalId: 1,
+					Voter:      myAddr.String(),
+					Options: []*v1.WeightedVoteOption{
+						{Option: v1.OptionYes, Weight: sdk.NewDecWithPrec(49, 2).String()},
+						{Option: v1.OptionNo, Weight: sdk.NewDecWithPrec(5, 1).String()},
```

Comment:
This silently changes the backend from x/gov v1beta1 to x/gov v1. Is this safe? Can it create any undesired behaviour change for contracts designed for v1beta1? How do we know which governance module chains want to use?

My approach so far was to only support the classic v1beta1 and treat v1 as a new unsupported module. But generally I am extremely unsure about the abstraction level of those CosmWasm message types.

Reply:
The gov v1beta [impl](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0-rc1/x/gov/keeper/msg_server.go#L245 ) is an adapter to v1. To be on the safe side, I had an [e2e test](https://github.com/CosmWasm/wasmd/pull/1149/files#diff-da54d221059499f9e50b24d5bdc84477815b5f6db37f1f6547ae85035d7bfb03) introduced with the sdk47 develop branch to cover votes. Just noticed that they don't include weighted votes yet but classic ones